synapse/util/caches/stream_change_cache.py

```diff
@@ -213,16 +213,17 @@ def has_any_entity_changed(self, stream_pos: int) -> bool:
         """
         assert isinstance(stream_pos, int)
 
-        if not self._cache:
```

Comment:
Note that `has_any_entity_changed` only has a single caller: `get_partial_current_state_deltas`, which is called from `get_current_state_deltas` which affects presence, stats, and the user directory.

Reply:
Searching the matrix.org database to see if there are descrepencies that could be due to this (or #14435 or #14592) shows that there are rooms where the room stats do not match current state events, this makes me think the user directory table might also be out of date? (I'll file a separate issue about this, I think.)

```sql
SELECT room_id, event_id, name AS stats_name, json::json->'content'->>'name' AS json_name
FROM current_state_events
INNER JOIN event_json USING (event_id, room_id)
INNER JOIN room_stats_state USING (room_id)
WHERE
    type = 'm.room.name' AND state_key = ''
    AND name != json::json->'content'->>'name';
```

Reply:
See #14643.